### DISCUSSION TITLE
"RuntimeException: Unable to select a random city id." When altering the demographics.csv file

### DISCUSSION THREAD

**Discussion by mikad98** [synthetichealth/synthea#1127] [Q&A] [answered]
I'm trying to alter the demographics.csv file such as that there are only 16 entries for each federal state of Germany.

Despite me adhering to the stated "Demographics File Format" in the Synthea Wiki, I keep getting the RuntimeException Error "Unable to select a random city id." when trying to execute "run_synthea".

I played around a bit with the original demographics.csv file and I found that if I cut the file to the first 10.000 rows, I do get the same error. However, if I cut the demographics.csv file to the first 30.000 rows and then execute "run_synthea" everything works perfectly fine. How is that possible?

Thanks for your help!

**Comment by jawalonoski** [comment]
I don't know what to say about varying your number of rows. The number of rows shouldn't matter, unless there are data quality issues that are causing the code to fail. Try debugging it.

That specific error `"Unable to select a random city id."` is related to selection of a random city. It requires that the total population of all cities be non-zero and each city have a non-zero population. 

Every city needs a unique `ID` in the file and every city needs to have a non-zero `POPESTIMATE2015`.

Check your file.

**Reply by mikad98** [reply]
FYI, I tried debugging the code, and here is what I found:

For me, the "randomCityId" method (in the Location.java file) only works, if there are entries in the demographics.csv file with STNAME "Massachusetts". If there is not (this was the case when I cut the demographics.csv file to the first 10.000 rows) the for-loop does not get executed, therefore throwing the RuntimeException "Unable to select a random city Id". This also explains why run_synthea perfectly worked, when I cut the file to the first 30.000 rows.

**Reply by mikad98** [reply]
I also noticed that whenever I execute "run_synthea", only patients from Massachusetts are generated, even if I pass "run_synthea -p 100" for example. This does only change if I execute e.g. "run_synthea Alabama". Then, also patients from other states (Alabama in this case) are generated.

**Reply by jawalonoski** [reply]
@mikad98 In the README file

```
run_synthea [-s seed] [-p populationSize] [state [city]]
```
or from the `run_synthea` help/usage instructions:

```
Options: [-s seed] [-cs clinicianSeed] [-p populationSize]
         [-r referenceDate as YYYYMMDD]
         [-e endDate as YYYYMMDD]
         [-g gender] [-a minAge-maxAge]
         [-o overflowPopulation]
         [-c localConfigFilePath]
         [-d localModulesDirPath]
         [-i initialPopulationSnapshotPath]
         [-u updatedPopulationSnapshotPath]
         [-t updateTimePeriodInDays]
         [-f fixedRecordPath]
         [-k keepMatchingPatientsPath]
         [--config*=value]
          * any setting from src/main/resources/synthea.properties
Examples:
run_synthea Massachusetts
run_synthea Alaska Juneau
run_synthea -s 12345
run_synthea -p 1000
run_synthea -s 987 Washington Seattle
run_synthea -s 21 -p 100 Utah "Salt Lake City"
run_synthea -g M -a 60-65
run_synthea -p 10 --exporter.fhir.export=true
run_synthea --exporter.baseDirectory="./output_tx/" Texas
```

If you don't specify a state or city, Synthea defaults to Massachusetts. If you don't have data for Massachusetts, it won't work. If you have data for somewhere else, you have to specify the location as a command-line argument (in order to override the default value of Massachusetts). If you only specify a hundred patients, it is not going to randomly pick them from across the United States (e.g., Alabama), it uses the default location.

**Comment by mikad98** [comment, marked as answer]
Could you maybe provide me a short "how to", for my task? Respectively a brief explanation of how you'd approach my problem?

Specifically, I want to create a patient population for Germany. I have a given patient population distribution regarding the 16 federal states of Germany. For example, I aim for a synthesized dataset where 5% of the population are from the federal state "lower saxony", 3% of the people from the federal state "Bremen" etc. The cities, zip codes etc. are not important for me, I just have the aim of a specific distribution regarding these federal states. 

Originally, my approach was to just replace the demographics.csv file with a new one that has 16 rows, each row for a respective federal state, such like here (only the fist 2 rows of the dataframe shown here):
![image](https://user-images.githubusercontent.com/75733150/186904184-b53a3ce4-323c-4dae-81fc-72759f2ac36b.png)

However, this approach leads to the problems mentioned above. Do I miss something? Your help would be very much appreciated!

**Reply by jawalonoski** [reply]
Have you read the series wiki articles about https://github.com/synthetichealth/synthea/wiki/Other-Areas? This is essentially a HOW-TO for your exact problem.

Another option is to look at the Germany example in the Synthea International repo (these examples are contributed by the community and I have not personally tested each one, so I can't guarantee they work) https://github.com/synthetichealth/synthea-international/tree/master/de

You cannot just replace the demographics file. You MUST also replace timezones, zip codes, payers, and providers (after all, German people won't buy American Health Insurance or fly to states for care). The program will not work without a consistent set of data. If you don't care about those things, the data does not have to be accurate, but it must be there.

**Reply by mikad98** [reply]
Thank you, that really helped a lot!

I somehow got stuck at the Wiki-Article "Demographics for Other Areas", where it says "If you modify or replace the demographics file, **_you may also need_** to replace the "Zip or Postal Codes" file..."

Therefore I thought that it isn't mandatory to also replace timezones, zip codes etc.

**Reply by jawalonoski** [reply]
OK, I'll fix that article. Thank you for pointing that out.
